docs/DesignDocs/SealingAPI.md

```diff
@@ -0,0 +1,591 @@
+
+    OE_CHECK(oe_get_seal_key(
+        (uint8_t*)&header->key_info,
+        sizeof(header->key_info),
+        &key,
+        &key_size));
+
+    OE_CHECK(oe_aes_gcm_decrypt(
+        key,
+        key_size,
+        header->iv,
+        sizeof(header->iv),
+        payload + header->ciphertext_size,
+        header->payload_size - header->ciphertext_size,
+        payload,
+        header->ciphertext_size,
+        header->tag));
+
+    if (plaintext)
+        *plaintext = payload;
+    if (plaintext_size)
+        *plaintext_size = header->ciphertext_size;
+    if (additional_data)
+        *additional_data = payload + header->ciphertext_size;
+    if (additional_data_size)
+        *additional_data_size = header->payload_size - header->ciphertext_size;
+
+done:
+    oe_free_seal_key(key, NULL);
+    return result;
+}
+```
+
+### oe_get_seal_key_info()
+
+```C
+/**
+ * Get enclave-specific key information that could be passed to
+ * oe_get_seal_key_v2() to derive a symmetric encryption key coupled to the
+ * enclave platform.
+ *
+ * @param[in] seal_policy The policy for the identity properties used to derive
+ * the seal key.
+ * @param[in] entropy Optional parameter containing additional entropy for key
+ * derivation. If \c NULL, oe_get_seal_key_info() will generate entropy on
+ * behalf of the caller.
+ * @param[in] entropy_size Must be \c 0 if \p entropy is \c NULL.
+ * @param[in] tee_specifc \c OE_SEAL_TEE_AGNOSTIC to use TEE defaults. If \c
+ * OE_SEAL_SGX is specified, the rest of \p tee_specifc is interpreted as the
+ * \a attribute_mask for \c EGETKEY on SGX.
+ * @param[out] key_info On success this points to the enclave-specific key
+ * information which can be used to retrieve the key by passing it to
+ * oe_get_seal_key_v2(). Freed by calling oe_free_seal_key() or oe_free_key().
+ * @param[out] key_info_size On success, this is the size of the \p key_info
+ * buffer.
+ *
+ * @retval OE_OK The seal key info was successfully requested.
+ * @retval OE_INVALID_PARAMETER At least one parameter is invalid.
+ * @retval OE_OUT_OF_MEMORY Failed to allocate memory.
+ */
+oe_result_t oe_get_seal_key_info(
+    oe_seal_policy_t seal_policy,
+    uint8_t* entropy,
+    size_t entropy_size,
+    uint64_t tee_specifc,
+    uint8_t** key_info,
+    size_t* key_info_size);
+```
+
+`oe_get_seal_key_info()` duplicates largely the logic of
+`oe_get_seal_key_by_policy()`. Please see [User Experience](#user-experience)
+earlier in this documentation for an explanation on the intention of this
+API.
+
+`seal_policy` selects the type of enclave identity and has the same definition
+and supported values as in `oe_get_seal_key_by_policy()`.
+
+`entropy`/`entropy_size` supplies additional data as input to the KDF.
+Different TEEs may expect different entropy sizes, e.g., SGX expects 256-bit
+entropy. If `entropy_size` isn't the same as expected, `entropy` will be hashed
+before being passed on to the underlying TEE. `entropy` could be `NULL` and if
```

Comment:
I don't understand this.  I'd expect it to fail if you give it insufficient entropy.

Reply:
Each TEE has entropy built in (e.g., SGX has fuse keys). This **extra** entropy is optionally. Per NIST.SP800-108, inputs to a *KDF* may include a *Label* and/or a *Context*. The latter may contain extra entropy, hence the name of this parameter.